filebeat/module/system/auth/config/auth.yml

```diff
@@ -15,3 +46,25 @@ processors:
 tags: {{ .tags | tojson }}
 
 publisher_pipeline.disable_host: {{ inList .tags "forwarded" }}
+
+{{ end }}
+
+{{ $journaldDetected := false }}
+
+{{if eq .builtin.osFamily "debian" }}
+  {{if eq .builtin.osVersion "12 (bookworm)" }}
+    {{ $journaldDetected = true }}
```

Comment:
I don't think we should be matching on OS, I think we need to be matching on some kind of capability.  I'd much rather see something where we try to detect if the data we need for the system auth module is in files or journald and then act accordingly.  I don't want to maintain a list of osFamily & osVersions, also it means that things like https://www.devuan.org/ get interesting as well, or other Distros where you can choose your init system.

Reply:
The first "assignment" on this task was to define what will be the best option to choose.
I'm sure he tried to have something that won't require us to maintain a list of osFamily/osVersion to define which input should be used but I will let @belimawr keeping me honest here.

Reply:
I agree with you Lee, hence I implemented the `use_journald` and `use_logs` option to force wither solution.

I did some research and did not find an assertive way to identify if a system is sending all logs we expect (syslog and auth in this context) to journald. Ubuntu is an interesting example, their [docs](https://ubuntu.com/tutorials/viewing-and-monitoring-log-files#2-log-files-locations) talks about the log files, systemd/jouranld are there and we can ingest them using the journald input.

Looking if the default files are there seems fragile, Debian 12 [release notes](https://www.debian.org/releases/bookworm/amd64/release-notes/ch-information.en.html#changes-to-system-logging) mention that `rsyslog` can be installed to have the traditional log files back, so if an user installs it for testing or what ever reason, the files are created, they remove it and install Filebeat + system integration. We might detect the files with data, but it will not be complete and new data won't be added there.

If we focus only on the supported Linux distributions, the list is not that long.

I'm open for suggestions on how to detect where the system logs are, if you have any idea or direction I could go looking for a better way to solve this, I'm open for it.

Reply:
What if you just try to ingest from both source by default? Most of the time, one of the inputs will be doing nothing.

This also has the advantage of never not collecting logs, it will only ever duplicate them, and at that point we can require the user to disable the one they don't want.

Reply:
Also worth looking at how other log collection agents solve this problem for inspiration.

Reply:
Always collecting both by default being viable potentially depends on how Debian 11 sets up logging, does it have both syslog and journald? I would not want to surprise existing users with working configurations with data duplication.

Putting it distro specific checks will require constant (but infrequent) maintenance and may break users that are happily functioning on officially unsupported distros.

Reply:
> Putting in distro specific checks will require constant (but infrequent) maintenance (...)

I know, and while not my proffered choice, that seemed to be the most reliable way to ensure we're collecting the logs from the correct place.

> (...) and may break users that are happily functioning on officially unsupported distros.

How would that break users that are functioning on officially unsupported distros? Like if they're on Debian 12 with rsyslog + system module and we release support+default for Journald?

That would duplicate data, I would not call that "breaking", or is it another scenario I'm not seeing? Either way, the users can always manually choose if they want to use journald or traditional log files.


> What if you just try to ingest from both source by default? Most of the time, one of the inputs will be doing nothing.

That will duplicate data in many systems/distros. Ubuntu being one of them that uses both, even though it only documents the traditional log files (see https://github.com/elastic/beats/pull/41061#discussion_r1787880734), Debian 11, according to the [release notes of Debain 12](https://www.debian.org/releases/bookworm/amd64/release-notes/ch-information.en.html#changes-to-system-logging) also uses both:
> In bullseye, rsyslog was installed by default and the systemd journal was configured to forward log messages to rsyslog, which writes messages into various text files such as /var/log/syslog.

I also tested Debian 10 and Ubuntu 22.04 in both distros I found the same messages on both: journald and the traditional log files. I don't have any documentation to back this up, however I believe that all distros that use systemd have the system logs in traditional log files and journald, hence if we ran both inputs we'd duplicate data in all distros that have not stopped using traditional log files like Debian 12 did.

Reply:
This might be a crazy idea.  But what if we made a new input.  The input is smart enough to peek at log sources, but doesn't update any registry settings.  The input has some decision tree it uses to pick either filestream or journalctl, this could be overridden with a user supplied option.  It then makes the config and runs the real input.

Some advantages to this approach:

1. We can log why we went filestream or journalctl, so users know why something is happening
3. The decision tree can be based on presence of log sources as well the contents of the log sources.  We could even use things like os family and os version if absolutely necessary.
4. Users can override the decision tree
5. The existing inputs are doing the real work, we are just making configs for them
6. This avoids data duplication

Reply:
That is an interesting idea @leehinman, somehow it will be similar to the `container` input that configures and starts a log input, however in this case we'd be choosing between the `journald` and `log` inputs (the current system module uses the `log` input, not `filestream`).

That would effectively migrate the logic I implemented in the template to Go code, the decision tree could be something very simple:
- If either `use_journald` or `use_logs` is true, we use what the user set.
- If both are set to true, we log an error and do not start any input
- If none of the files in `var.paths` exist, we use `journald`
- If none of the above selected an input, we use the `log` input, effectively, rendering the same config we currently render for the system module.

I'd also not document this input and try to prevent it from being defined in `filebeat.inputs` (I'm not sure if that is possible, but I'll try)

@cmacknz, if you agree I'll proceed with this implementation.

Reply:
I like the idea, it's clear we need some conditional logic and doing it using templates seems not feasible.

How would we start using the new input without breaking existing users? Would it have to be able to use the log input, filestream input, or journald input based on context?

I would argue it should probable default to using either journald or filestream, and only replace filestream with log based on a parameter like `use_depreacted_log_input` or something. 

This would avoid net new uses of this input from using the log input, and let us use the log input in places where we have to for backwards compatibility like in the system integration/module today.

Reply:
> How would we start using the new input without breaking existing users? Would it have to be able to use the log input, filestream input, or journald input based on context?

We need this for filebeat system module and for system integration.  So I think we could "just" use the input in those places and as long as we maintain the id that the registry uses to find offset it should be transparent to the users.

Reply:
Makes sense, I don't think there's a way to discourage use of this input since it needs to be GA for us to use it in the system module, so we probably still need something along the lines of a `use_depreacted_log_input` to ensure people using it by default do not use the log input by default.

P.S. it is likely for 9.x we are going to force every use of the log input to opt in to using it with a `use_deprecated_input: true` flag in the same way, as a way of forcing users to discover the log input is deprecated.

Reply:
> Makes sense, I don't think there's a way to discourage use of this input since it needs to be GA for us to use it in the system module, so we probably still need something along the lines of a `use_depreacted_log_input` to ensure people using it by default do not use the log input by default.

I haven't thought deeply about this, but I think having this input opens up new possibilities for migrating users from log to filestream input.  Since we are peeking at log sources, we could also peek at registry entries to see if a path was previously ingested by log input.  That should give us more info and options when trying to convert from log input to filestream input.  Needs more thought to see if it really provides enough info/options.

Reply:
Taking a step back folks, what is the main focus of this new input? After reading all the comments it looks like you folks have a different view than what's in my mind.

My idea for this new input (let's call it `system-logs` for now) is to have it only for the system module in Filebeat. As I mentioned before, I'd even try to prevent users from manually configuring it (I'm not sure it's possible, but I'd try), keeping it as an implementation detail from the system module rather than an wide-open new feature.

I'm not following very closely how logging on Linux is changing from traditional log files to jouranld, but if this is a thing that will happen in more distros and applications, than I could see an interesting use case for this input aside the system module. Anyways, we can always start with the system-module use case and then expand/open up its usage.

That said, and focusing on the system module use case, I'd just not add  try to use the filestream input for now, let's keep it simple and focus on journald x log files (log input).

The migration from log input to filestream input is a solved problem, we have the take over mode, it is proven for any case aside from autodiscover, which is not a concern here. Any case where users cannot configure IDs/do not have much flexibility to craft a complex input configuration, migrating from log to filestream input is very easy, the system module is one of those cases. The only thing the user can configure is the list of globs to ingest, which makes. The only challenge here is the fact that modules can be reloaded, but we can focus on that later.

**Replying to your comments**

> How would we start using the new input without breaking existing users? Would it have to be able to use the log input, filestream input, or journald input based on context?

As long as we instantiate a log input with the same configurations as the current system module, there will be no impact for existing users, that's very simple to do, I don't see any challenge here. The log input has a global state for all of its instances, so as long as it's ingesting the same files, the state is not lost (it does not matter when/how the input is started), also we can ensure it will generate the same log input configuration as the system module currently generates. 

As I mentioned before, I'd not include filestream at this moment, we can always add it later as an improvement.

> I would argue it should probably default to using either journald or filestream, and only replace filestream with log based on a parameter like use_depreacted_log_input or something.
>
> This would avoid net new uses of this input from using the log input, and let us use the log input in places where we have to for backwards compatibility like in the system integration/module today.

To keep things simple I'd just not document this input for now, I'd consider it an implementation detail of the system module rather than a new feature we're adding to Filebeat.

> I haven't thought deeply about this, but I think having this input opens up new possibilities for migrating users from log to filestream input. Since we are peeking at log sources, we could also peek at registry entries to see if a path was previously ingested by log input. That should give us more info and options when trying to convert from log input to filestream input. Needs more thought to see if it really provides enough info/options.

That mostly sounds like what the take over mode already does, I do not believe we need a new input or a  new way to migrate users from the log input to filestream. I believe for that we just need to cover the cases where the take over mode does not work yet and maybe give an automated way to convert the configuration from the log input to a valid-and-ready-to-use filestream configuration.

## My plan after this discussion
I'll implement this new input (likely calling it system-logs):
- it will be an implementation detail of the system module for now (aka, no public docs),
- it will only choose between journald/log input (keeping the current behaviour)
- I'll use the logic I described on https://github.com/elastic/beats/pull/41061#discussion_r1790347566
- I'll document the new behaviour in the changelog.

Reply:
Overall your plan looks good, just adding some wider context below.

> As I mentioned before, I’d even try to prevent users from manually configuring it (I’m not sure it’s possible, but I’d try), keeping it as an implementation detail from the system module rather than an wide-open new feature.

This is not possible in standalone Beats. If the system module can use it, anyone can put it into their configuration manually.

> I’m not following very closely how logging on Linux is changing from traditional log files to jouranld, but if this is a thing that will happen in more distros and applications, than I could see an interesting use case for this input aside the system module.

Syslog is no longer installed by default. All logs are going into journald. This is happening at a different rate across many distros. Amazon Linux 2023 is not officially Debian, but it is Debian like, and is doing the same things.

> That said, and focusing on the system module use case, I’d just not add try to use the filestream input for now, let’s keep it simple and focus on journald x log files (log input).

If you don’t put filestream in the input, people who choose to use this input themselves will accidentally use the deprecated log input. There has to be a path to filestream in everything we do unfortunately. We can’t push new users to deprecated things. You could handle this as follow up work as it isn’t strictly required for the system module, but it has to be done.

> To keep things simple I’d just not document this input for now, I’d consider it an implementation detail of the system module rather than a new feature we’re adding to Filebeat.

👍

Reply:
> If you don’t put filestream in the input, people who choose to use this input themselves will accidentally use the deprecated log input. There has to be a path to filestream in everything we do unfortunately. We can’t push new users to deprecated things. You could handle this as follow up work as it isn’t strictly required for the system module, but it has to be done.

I agree, I'm trying to keep the changes that need to go into the FF as small as possible, we can always iterate and improve later.